contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,329 @@
+```
+should be changed to
+```go
+klog.Infof("delete object", ...)
+```
+
+Some logs are constructed solely from string formats. In those cases a message needs to be derived from context of
+log call.
+
+For example http access logs
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("%s %s: (%v) %v%v%v [%s %s]", ...)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("HTTP", ...)
+}
+```
+
+### Name arguments
+
+Even thou new structured logging functions have very similar function prototype `func (string, ...interface{}) it's
+has different meaning for variadic arguments. Instead just passing arguments, now we are passing key value pairs of
+argument name and argument value. This means when migrating log call we need to add additional string before each
+argument, that will be used as it's name.
+
+For example
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("Got request, path: %s, method: %s", r.Path, r.Method)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("Got request", "path", r.Path, "method", r.Method)
+}
+```
+
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different logs lines on `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
+names: `deployment`, `pod`, `node`, `replicaSet`.
+
+In situations when we want to log value of same meaning twice (e.g. transition between state) it is ok to name
+additional argument based on context, but leaving one main value with name unchanged.
+
+Example:
+
+```go
+func LogStatus(oldStatus, desiredStatus, newStatus string) {
+   klog.InfoS("Status changed", "status", newStatus, "oldStatus", oldStatus, "desiredStatus", desiredStatus)
+}
+```
+
+[lowerCamelCase]: https://en.wiktionary.org/wiki/lowerCamelCase
+
+### Use `klog.KObj` and `klog.KRef` for Kubernetes objects
+
+As part of structured logging migration we want to ensure that kubernetes objects references are consistent within
+codebase. Two new utility functions were introduced to klog `klog.KObj` and `klog.KRef`. Any reference
+(name, uid, namespace) to Kubernetes Object (Pod, Node, Deployment, CRD) should be rewritten to utilize those functions.
+
+For example
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("updated pod %s in namespace %s", pod.Name, pod.Namespace)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("updated pod", "pod", klog.KObj(pod))
+}
+```
+And
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("updated pod, uid: %s", pod.Uid)
```

Comment:
ObjectRef currently does not include uid so if we will replace uid with klog.KObj() we will have a regression when it comes to the information available in logs.

Reply:
As UID and namespace name pair are both unique identifiers of object I would not treat this change as regression as it preserves uniqueness.

There is also case when name/namespaces are not available, then usage of UID is ok.

Reply:
As I understand uid has a benefit of being unique across the time. So if a new resource will be created with the same name it will get a different uid.
I am not sure if there are any real use cases when this difference is relevant but I believe that this might be a regression for some users.

Reply:
Added more context in https://github.com/kubernetes/community/pull/4793#discussion_r431364645